openedx_tagging/core/tagging/models.py

```diff
@@ -627,3 +649,240 @@ def copy(self, object_tag: "ObjectTag") -> "ObjectTag":
+
+        Subclasses can override this method to get the value from their own models.
+        """
+        return str(instance.pk)
+    
+    def _get_instance(self, pk):
+        """
+        Gets the instance from `pk`
+        """
+        try:
+            return self.tag_class_model.objects.get(pk=pk)
+        except self.tag_class_model.DoesNotExist:
+            return None
+    
+    def _resync_tag(self, tag: Tag) -> Tag:
+        """
+        Resync the tag value with the value from the instance
+        """
+        instance = self._get_instance(tag.external_id)
+        if instance:
+            value = self._get_value_from_instance(instance)
+            if tag.value != value:
+                tag.value = value
+                tag.save()
+        return tag
+
+    def _set_tag(
+        self,
+        tag_ref,
+        object_tag: ObjectTag
+    ) -> ObjectTag:
+        """
+        Set or create the respective Tag on `object_tag`
+
+        This function is used on `tag_object`.
+        For this to work it is necessary that `tag_ref` 
+        is always the `pk` of the model to which you want
+        to associate it.
+        """
+        try:
+            tag = self.tag_set.get(
+                external_id=tag_ref,
+            )
+            # Run a resync of the tag
+            tag = self._resync_tag(tag)
+            object_tag.tag = tag
+        except (ValueError, Tag.DoesNotExist):
+            # Creates a new tag with the instance
+            instance = self._get_instance(tag_ref)
+            if not instance:
+                return None
+            new_tag = Tag(
+                taxonomy=self,
+                value=self._get_value_from_instance(instance),
+                external_id=self._get_external_id_from_instance(instance),
+            )
+            new_tag.save()
+            object_tag.tag = new_tag
+
+        return object_tag
+
+class UserSystemDefinedTaxonomy(ModelSystemDefinedTaxonomy):
+    """
+    User based system taxonomy class.
+    """
+
+    class Meta:
+        proxy = True
+
+    @property
+    def tag_class_model(self) -> Type:
+        """
+        Associate the user model
+        """
+        return get_user_model()
+
+    def _get_value_from_instance(
+        self,
+        instance
+    ) -> str:
+        """
+        Returns the username as tag value
+        """
+        return instance.get_username()
+
+
+class LanguageTaxonomy(SystemDefinedTaxonomy):
+    """
+    Language System-defined taxonomy
+
+    The tags are filtered and validated taking into account the 
+    languages available in Django LANGUAGES settings var
+    """
+
+    class Meta:
+        proxy = True
+
+    def _get_tags_query_set(self) -> models.QuerySet:
+        """
+        Returns a query set of available languages tags.
+        """
+        available_langs = self._get_available_languages()
+        return self.tag_set.filter(external_id__in=available_langs)
```

Comment:
Since LanguageTaxonomy is a subclass of Taxonomy, I think it's cleaner to [do this instead](https://github.com/open-craft/openedx-learning/pull/6/commits/bc2c31788d67b842cd26121a9dc67e17633ef18c).

Reply:
Added to the branch